### PR TITLE
Implement dutch auc

### PR DESCRIPTION
**Pull request by UladAleph**
Implemented contract for selling (and buying) an asset using a Dutch auction.

### CODE REVIEW

**Review by DamianStraszak — APPROVED** [review]

File: lib.rs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    }
+
+    impl From<PSP22Error> for Error {
+        fn from(inner: PSP22Error) -> Self {
+            Error::PSP22TokenCall(inner)
+        }
+    }
+
+    impl DutchAuction {
+        /// Constructor that initializes the contract storrage.
```
Comment (ipapandinas) [inline lib.rs]: Small typo: storage

File: lib.rs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        #[ink(message)]
+        pub fn min_price(&self) -> u128 {
+            self.min_price
+        }
+
+        /// Update the minimal price.
+        ///
+        /// Requires auction_owner to execute.
+        #[ink(message)]
+        pub fn set_min_price(&mut self, value: u128) -> Result<(), Error> {
```
Comment (ipapandinas) [inline lib.rs:149]: Not sure about having `set_min_price` and `set_end_time` once an auction has started; it depends on specs but we probably want to keep it permissionless.

**Review by ipapandinas — APPROVED** [review]

File: lib.rs (hunk 1 above)
Comment (UladAleph) [reply lib.rs]: fixed

File: lib.rs (hunk 2 above)
Comment (UladAleph) [reply lib.rs:149]: These are essential parameters for calculating the price and controlling the duration of auction. I believe it is good to leave a window for an owner to change them for some flexibility. Leaving them permissionless would introduce security risks. For example, if anybody can set the `min_price` it is possible they will then buy at a price lower then intended by the owner

**Pull request merged**